README.md
Simple Console
==============

Provides a simple, but useful cross-browser-compatible `console` logger.

[![Build Status][trav_img]][trav_site]
[![Coverage Status][cov_img]][cov_site]

[![Sauce Test Status][sauce_img]][sauce_site]

### Features

* Proxies to native functionality wherever possible.
* Enables `.apply` and `.bind` usage with `console.OPERATION`.
* Buildable from straight CommonJS source.
* Available as minified distributions as well.

### Installation

Install via [npm](https://www.npmjs.com/package/simple-console):

```
$ npm install simple-console
```

or [bower](http://bower.io/search/?q=simple-console):

```
$ bower install simple-console
```

### Usage

Import the `SimpleConsole` class into your code (via AMD, CommonJS, etc) and
use as a drop-in replacement for `console.OPERATION`. Creating a new object
does _not_ effect the existing `window.console` object.

#### AMD

```js
define(["simple-console"], function (SimpleConsole) {
  var con = new SimpleConsole();
  con.log("Hello world!");
});
```

#### CommonJS

```js
var SimpleConsole = require("simple-console");
var con = new SimpleConsole();
con.log("Hello world!");
```

#### VanillaJS

In your HTML:

```html
<!-- Option One: Minified -->
<script src="PATH/TO/simple-console/dist/simple-console.min.js"></script>

<!-- Option Two: Raw source -->
<script src="PATH/TO/simple-console/simple-console.js"></script>
```

In your JS:

```js
var con = new window.SimpleConsole();
con.log("Hello world!");
```

### Noop Logger

There are some cases where you will want to conditionally silence the logger.
You can do this by passing setting the `noop` option:

```js
var con = new SimpleConsole({ noop: true });
con.log("Hello world!"); // => Should _not_ output anything.
```

This is usually useful in a case where you build different capabilities based
on some external information, for example, React-style, this could be
something like:

```js
var con = new SimpleConsole({
  noop: process.env.NODE_ENV === "production"
});

con.log("Hello world!"); // => Should _not_ output anything in `production`.
```

### Polyfill

If you are looking to **polyfill** `console`, then you can:

```js
new SimpleConsole({ patch: true });
```

This **will** mutate the `window.console` object in ways that are not easily
undone, so consider this a "one way" patch. Moreover, if `window.console`
does not exist, it will create the object.

You can even go further and patch _and_ noop the logger with:

```js
new SimpleConsole({ patch: true, noop: true });
```

which ensures that _nothing_ logs anything.

**Note**: In addition to filling/patching missing behavior, the polyfill will
replace behavior that already _exists_ and _works_. This is presently due to
complexities likely normalizing and ensuring things like `.bind`, `.apply`
and `.call` work on all log methods.


### Development

Run checks, then development server:

```
$ gulp
```

Separated:

```
$ gulp check
$ gulp dev
```

Execute the demo code:

```
$ node examples/demo.js
$ phantomjs examples/demo.js
```

Navigations:

* **Test Page**: http://127.0.0.1:4321/test/test.html
* **Demo Page**: http://127.0.0.1:4321/examples/demo.html

Release process:

* Bump versions in `package.json`, `bower.json`
* Run `npm run-script everything`
* Edit `HISTORY.md`
* Run `gulp check:all`

### Also See

Similar projects that can help with `console`:

* [`console-polyfill`](https://github.com/paulmillr/console-polyfill)
* [`console.log-wrapper`](https://github.com/patik/console.log-wrapper)

### License
Copyright 2015 Formidable Labs, Inc.
Released under the [MIT](LICENSE.txt) License,

[trav]: https://travis-ci.org/
[trav_img]: https://api.travis-ci.org/FormidableLabs/simple-console.svg
[trav_site]: https://travis-ci.org/FormidableLabs/simple-console
[cov]: https://coveralls.io
[cov_img]: https://img.shields.io/coveralls/FormidableLabs/simple-console.svg
[cov_site]: https://coveralls.io/r/FormidableLabs/simple-console
[sauce]: https://saucelabs.com
[sauce_img]: https://saucelabs.com/browser-matrix/simple-console.svg
[sauce_site]: https://saucelabs.com/u/simple-console


HISTORY.md
History
=======

## 0.1.1

* Make Node.js compatible.
* Aggregate test coverage results for different browsers / server.

## 0.1.0

* Refactor `SimpleConsole` to take options object with explicit `patch` and
  `noop` options.
* Remove `SimpleConsole.patch()` and simplify polyfill / noop instructions.

## 0.0.4

* Add missing `dist` file.

## 0.0.3

* **WARNING**: Missed `dist` file. **Do not use this version**.
* Add props/method to `this` as well as target context.
* Add explicit documentation for sinkhole logger and use cases.

## 0.0.2

* Should have full functionality as instance logger / global polyfill.
* Add full tests for Chrome, Safari, IE, Firefox, etc.
* Add polyfill method `SimpleConsole.patch()`.
* Add basic unit tests.

## 0.0.1

* Initial Release.


test/spec/simple-console.spec.js
var isNode = typeof window === "undefined";
var root = isNode ? global : window;

describe("simple-console", function () {

  describe("no real console logger", function () {
    // Track if bind is not available for certain platforms.
    var noBind = !Function.prototype.bind,
      conStub,
      con;

    beforeEach(function () {
      // By default, returns `undefined`.
      conStub = sinon.stub(SimpleConsole.prototype, "_getConsole");
      con = new SimpleConsole();
    });

    afterEach(function () {
      conStub.restore();
    });

    it("should invoke lots of functions and not die", function () {
      con.log("log");
      con.log.apply(con, ["log apply"]);
      con.log.call(con, "log call");
      con.warn("warn");
      con.warn.apply(con, ["warn apply"]);
      con.warn.call(con, "warn call");
      con.error("error");
      con.error.apply(con, ["error apply"]);
      con.error.call(con, "error call");
    });

    it("should handle swap between warn or log", function () {
      // Short-circuit: need bind.
      if (noBind) { return; }

      // This is a very common use case for logging.
      // See, e.g.: https://github.com/FormidableLabs/biff/issues/3
      var warn = (con.warn || con.log).bind(console);

      warn("warn");
      warn.apply(con, ["warn apply"]);
      warn.call(con, "warn call");
    });
  });

  describe("noop logger", function () {
    it("should create a noop logger", function () {
      var con = new SimpleConsole({
        noop: true
      });

      con.log("log");
      con.log.apply(con, ["log apply"]);
      con.log.call(con, "log call");
      con.warn("warn");
      con.warn.apply(con, ["warn apply"]);
      con.warn.call(con, "warn call");
      con.error("error");
      con.error.apply(con, ["error apply"]);
      con.error.call(con, "error call");
    });
  });

  // These specs will log out to console.
  // This is generally disfavored in tests, but we're _testing_ a logger
  // and want to have things blow up if anything is wrong.
  describe("actually log out", function () {
    describe("no function bind", function () {
      var bindStub,
        con;

      beforeEach(function () {
        // By default, returns `undefined`.
        bindStub = sinon.stub(SimpleConsole.prototype, "_getBind");
        con = new SimpleConsole();
      });

      afterEach(function () {
        bindStub.restore();
      });

      it("should invoke lots of functions and not die", function () {
        con.log("log");
        con.log.apply(con, ["log apply"]);
        con.log.call(con, "log call");
        con.warn("warn");
        con.warn.apply(con, ["warn apply"]);
        con.warn.call(con, "warn call");
        con.error("error");
        con.error.apply(con, ["error apply"]);
        con.error.call(con, "error call");
      });
    });

    describe("native functionality", function () {
      var con;

      beforeEach(function () {
        con = new SimpleConsole();
      });

      it("should invoke lots of functions and maybe log", function () {
        con.log("log");
        con.log.apply(con, ["log apply"]);
        con.log.call(con, "log call");
        con.warn("warn");
        con.warn.apply(con, ["warn apply"]);
        con.warn.call(con, "warn call");
        con.error("error");
        con.error.apply(con, ["error apply"]);
        con.error.call(con, "error call");
      });
    });

    // **NOTE**: This test should come last because we can't really _undo_
    // all the stuff we're doing...
    describe("monkey patch real console", function () {

      before(function () {
        /*eslint-disable no-new*/
        new SimpleConsole({
          patch: true
        });
        /*eslint-enable no-new*/
      });

      it("should invoke lots of functions and maybe log", function () {
        root.console.log("log");
        root.console.log.apply(root.console, ["log apply"]);
        root.console.log.call(root.console, "log call");
        root.console.warn("warn");
        root.console.warn.apply(root.console, ["warn apply"]);
        root.console.warn.call(root.console, "warn call");
        root.console.error("error");
        root.console.error.apply(root.console, ["error apply"]);
        root.console.error.call(root.console, "error call");
      });
    });

    // **NOTE**: `console` is pretty much unusable past this point because
    // we've now overwritten it with something that noop's all operations.
    describe("monkey patch and noop console", function () {
      // Skip if Node so Gulp, etc. still usable...
      if (isNode) { return; }

      before(function () {
        /*eslint-disable no-new*/
        new SimpleConsole({
          patch: true,
          noop: true
        });
        /*eslint-enable no-new*/
      });

      it("should invoke lots of functions and maybe log", function () {
        root.console.log("log");
        root.console.log.apply(root.console, ["log apply"]);
        root.console.log.call(root.console, "log call");
        root.console.warn("warn");
        root.console.warn.apply(root.console, ["warn apply"]);
        root.console.warn.call(root.console, "warn call");
        root.console.error("error");
        root.console.error.apply(root.console, ["error apply"]);
        root.console.error.call(root.console, "error call");
      });
    });
  });
});


simple-console.js
/*!
 * simple-console
 * --------------
 * A small, cross-browser-friendly `console` wrapper.
 */
(function () {
  /*eslint-disable consistent-this*/
  var root = this;
  /*eslint-enable consistent-this*/

  // Patches
  var EMPTY_OBJ = {};
  var NOOP = function () {};

  // Console attributes
  var props = ["memory"];
  var meths = (
    "assert,clear,count,debug,dir,dirxml,error,exception,group," +
    "groupCollapsed,groupEnd,info,log,markTimeline,profile,profiles," +
    "profileEnd,show,table,time,timeEnd,timeline,timelineEnd,timeStamp,trace" +
    ",warn"
  ).split(",");

  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/
  // Global_Objects/Array/isArray
  var isArray = Array.isArray || function (arg) {
    return Object.prototype.toString.call(arg) === "[object Array]";
  };

  /**
   * Console abstraction.
   *
   * Provides a drop-in replacement for `console` with a few extras.
   *
   * @param {Object} opts       options object
   * @param {Object} opts.patch patch underlying `console`? (default: `true`)
   * @param {Object} opts.noop  have all methods be NOOP?   (default: `true`)
   */
  var SimpleConsole = function (opts) {
    var self = this;
    var i;

    // Starting state.
    var bind = self._getBind();
    var con = self._getConsole();
    var noConsole = !con; // Stash if console unavailable.

    // Protect variables.
    opts = opts || {};
    con = con || {};

    // Target: Add properties to console if patching.
    var target = {};
    if (opts.patch) {
      // Ensure that `window.console` is actually created, and set as target.
      target = root.console = root.console || target;
    }

    // Patch properties, methods.
    for (i = 0; i < props.length; i++) {
      // *Note*: Could consider _copying_ values.
      self[props[i]] = target[props[i]] = opts.noop ?
        EMPTY_OBJ :
        con[props[i]] || EMPTY_OBJ;
    }

    // Enable "apply" and "bind" on methods by converting to real function.
    // See: http://patik.com/blog/complete-cross-browser-console-log/
    for (i = 0; i < meths.length; i++) {
      // Set context _and_ the target.
      self[meths[i]] = target[meths[i]] = (function (methFn) {
        if (opts.noop || noConsole || !methFn) {
          // Noop cases.
          return NOOP;

        } else if (isArray(methFn)) {
          // Straight assign any array objects.
          // *Note*: Could do `.slice(0);` to clone.
          //
          // Fixes Safari on Mac OS X 10.9 on Sauce.
          // Issue is `console.profiles`, which is an array.
          // See: https://github.com/FormidableLabs/simple-console/issues/3
          // See: https://saucelabs.com/tests/9a89e381c91c4e43b25ab8ee16a514e1
          return methFn;

        } else if (bind) {
          // IE9 and most others: Bind to our create real function.
          // Should work if `console.FOO` is `function` or `object`.
          return bind.call(methFn, con);
        }

        // IE8: No bind, so even more tortured.
        return function () {
          Function.prototype.call.call(methFn, con,
            Array.prototype.slice.call(arguments));
        };
      })(con[meths[i]]);
    }
  };

  /**
   * Accessor to console object. (Cached).
   *
   * @returns {Object} the console object or `null` if unavailable.
   * @api private
   */
  var _console;
  SimpleConsole.prototype._getConsole = function () {
    if (typeof _console !== "undefined") { return _console; }
    _console = root.console || null;
    return _console;
  };

  /**
   * Accessor to bind object. (Cached).
   *
   * @returns {Object} the console object or `null` if unavailable.
   * @api private
   */
  var _bind;
  SimpleConsole.prototype._getBind = function () {
    if (typeof _bind !== "undefined") { return _bind; }
    _bind = Function.prototype.bind || null;
    return _bind;
  };

  // UMD wrapper: Borrowed from webpack version.
  /* istanbul ignore next */
  /*global exports define*/
  if (typeof exports === "object" && typeof module === "object") {
    // CommonJS
    module.exports = SimpleConsole;
  } else if (typeof define === "function" && define.amd) {
    // AMD
    define(function () {
      return SimpleConsole;
    });
  } else {
    // VanillaJS / Old exports
    var mod = typeof exports === "object" ? exports : root;
    mod.SimpleConsole = SimpleConsole;
  }
})();


gulpfile.js
/**
 * Gulp file.
 */
var _ = require("lodash");
var gulp = require("gulp");
var eslint = require("gulp-eslint");
var jscs = require("gulp-jscs");
var karma = require("gulp-karma");
var mocha = require("gulp-mocha");
var istanbul = require("gulp-istanbul");
var connect = require("gulp-connect");

// ----------------------------------------------------------------------------
// EsLint
// ----------------------------------------------------------------------------
var _eslint = function (files, cfg) {
  return function () {
    return gulp
      .src(files)
      .pipe(eslint(cfg))
      .pipe(eslint.formatEach("stylish", process.stderr))
      .pipe(eslint.failOnError());
  };
};

gulp.task("eslint:frontend", _eslint([
  "simple-console.js"
], {
  envs: ["browser"]
}));

gulp.task("eslint:frontend:test", _eslint([
  "examples/**/*.js",
  "test/**/*.js"
], {
  envs: ["browser", "mocha"],
  globals: ["expect", "sinon", "SimpleConsole", "global"]
}));

gulp.task("eslint:backend", _eslint([
  "gulpfile.js",
  "webpack.*.js"
], {
  envs: ["node"]
}));

gulp.task("eslint", [
  "eslint:frontend", "eslint:frontend:test",
  "eslint:backend"
]);

// ----------------------------------------------------------------------------
// JsCs
// ----------------------------------------------------------------------------
gulp.task("jscs", function () {
  return gulp
    .src(["*.js"])
    .pipe(jscs());
});

// ----------------------------------------------------------------------------
// Publication checks
// ----------------------------------------------------------------------------
gulp.task("chk", function () {
  var bower = require("./bower.json");
  var npm = require("./package.json");

  if (bower.version !== npm.version) {
    throw new Error(
      "Bower (" + bower.version + ") / NPM  (" + npm.version + ") mismatch");
  }
});

// ----------------------------------------------------------------------------
// Test - Frontend
// ----------------------------------------------------------------------------
// Use `node_modules` Phantom
process.env.PHANTOMJS_BIN = "./node_modules/.bin/phantomjs";

// Sauce labs environments.
var SAUCE_ENVS = {
  /*eslint-disable camelcase*/
  // Already tested in Travis.
  // sl_firefox: {
  //   base: "SauceLabs",
  //   browserName: "firefox"
  // },
  sl_chrome: {
    base: "SauceLabs",
    browserName: "chrome"
  },
  sl_safari_6_8: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "OS X 10.8",
    version: "6.0"
  },
  sl_safari_7_9: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "OS X 10.9",
    version: "7.0"
  },
  sl_safari_8_10: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "OS X 10.10",
    version: "8.0"
  },
  sl_ie_8: {
    base: "SauceLabs",
    browserName: "internet explorer",
    platform: "Windows XP",
    version: "8"
  },
  sl_ie_9: {
    base: "SauceLabs",
    browserName: "internet explorer",
    platform: "Windows 7",
    version: "9"
  },
  sl_ie_10: {
    base: "SauceLabs",
    browserName: "internet explorer",
    platform: "Windows 7",
    version: "10"
  },
  sl_ie_11: {
    base: "SauceLabs",
    browserName: "internet explorer",
    platform: "Windows 7",
    version: "11"
  }
  /*eslint-enable camelcase*/
};

// SauceLabs tag.
var SAUCE_BRANCH = process.env.TRAVIS_BRANCH || "local";
var SAUCE_TAG = process.env.SAUCE_USERNAME + "@" + SAUCE_BRANCH;

// Test wrapper.
var testFrontend = function () {
  var files = [
    // Libraries
    "node_modules/expect.js/index.js",
    "node_modules/sinon/pkg/sinon.js",

    // App files
    "simple-console.js",

    // Tests
    "test/spec/simple-console.spec.js"
  ];

  var opts = _.extend.apply(_, [{
    frameworks: ["mocha"],
    port: 9999,
    reporters: ["spec", "coverage"],
    preprocessors: {
      "simple-console.js": ["coverage"]
    },
    coverageReporter: {
      reporters: [
        { type: "json", file: "coverage.json" },
        { type: "lcov" },
        { type: "text-summary" }
      ],
      dir: "coverage/client"
    },
    client: {
      captureConsole: true,
      mocha: {
        ui: "bdd"
      }
    }
  }].concat(_.toArray(arguments)));

  return function () {
    return gulp
      .src(files)
      .pipe(karma(opts))
      .on("error", function (err) {
        throw err;
      });
  };
};

gulp.task("test:frontend:dev", testFrontend({
  singleRun: true,
  browsers: ["PhantomJS"]
}));

gulp.task("test:frontend:ci", testFrontend({
  singleRun: true,
  browsers: ["PhantomJS", "Firefox"]
}));

gulp.task("test:frontend:sauce", testFrontend({
  singleRun: true,
  reporters: ["spec", "saucelabs", "coverage"],
  sauceLabs: {
    testName: "simple-console - Frontend Unit Tests",
    tags: [SAUCE_TAG],
    "public": "public"
  },
  // Timeouts: Allow "n" minutes before saying "good enough". See also:
  // https://github.com/angular/angular.js/blob/master/karma-shared.conf.js
  captureTimeout: 0, // Pass through to SL.
  customLaunchers: SAUCE_ENVS,
  browsers: Object.keys(SAUCE_ENVS)
}));

gulp.task("test:frontend:all", testFrontend({
  port: 9998,
  browsers: ["PhantomJS", "Firefox", "Chrome", "Safari"]
}));

// ----------------------------------------------------------------------------
// Test - Backend
// ----------------------------------------------------------------------------
gulp.task("test:backend", function (done) {
  // First, cover files.
  gulp
    .src(["simple-console.js"])
    .pipe(istanbul({
      includeUntested: true
    }))
    .pipe(istanbul.hookRequire())
    .on("finish", function () {
      // Node adapter.
      global.sinon = require("sinon");
      global.SimpleConsole = require("./simple-console");

      // Second, run the tests
      gulp
        .src(["test/**/*.js"], { read: false })
        .pipe(mocha({
          ui: "bdd",
          reporter: "spec"
        }))
        .on("error", function (err) {
          throw err;
        })
        .pipe(istanbul.writeReports({
          dir: "./coverage/server",
          reportOpts: { dir: "./coverage/server" },
          reporters: ["lcov", "json", "text-summary"]
        }))
        .on("end", done);
    });
});

// ----------------------------------------------------------------------------
// Servers
// ----------------------------------------------------------------------------
// Static server.
// Test page: http://127.0.0.1:4321/test/test.html
gulp.task("server", function () {
  connect.server({
    root: __dirname,
    port: 4321
  });
});

// ----------------------------------------------------------------------------
// Aggregations
// ----------------------------------------------------------------------------
gulp.task("style", ["chk", "eslint", "jscs"]);
gulp.task("check-base", ["style", "test:backend"]);
gulp.task("check", ["check-base", "test:frontend:dev"]);
gulp.task("check:ci", ["check-base", "test:frontend:ci"]);
gulp.task("check:all", ["check-base", "test:frontend:all"]);
gulp.task("dev", ["server"]);
gulp.task("default", ["check", "dev"]);
